samples/core/binaries/main.cpp

```diff
@@ -0,0 +1,218 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// OpenCL SDK includes
+#include <CL/Utils/Utils.hpp>
+#include <CL/SDK/Context.hpp>
+#include <CL/SDK/Options.hpp>
+#include <CL/SDK/CLI.hpp>
+
+// STL includes
+#include <climits>
+#include <iostream>
+#include <fstream>
+
+// TCLAP includes
+#include <tclap/CmdLine.h>
+
+// Sample-specific option
+struct BinariesOptions
+{
+    size_t start;
+    size_t length;
+};
+
+// Add option to CLI parsing SDK utility
+template <> auto cl::sdk::parse<BinariesOptions>()
+{
+
+    return std::make_tuple(
+        std::make_shared<TCLAP::ValueArg<size_t>>(
+            "s", "start", "Starting number", false, 1, "positive integral"),
+        std::make_shared<TCLAP::ValueArg<size_t>>("l", "length",
+                                                  "Length of input", false,
+                                                  100000, "positive integral"));
+}
+template <>
+BinariesOptions cl::sdk::comprehend<BinariesOptions>(
+    std::shared_ptr<TCLAP::ValueArg<size_t>> start_arg,
+    std::shared_ptr<TCLAP::ValueArg<size_t>> length_arg)
+{
+    return BinariesOptions{ start_arg->getValue(), length_arg->getValue() };
+}
+
+int main(int argc, char* argv[])
+{
+    try
+    {
+        // Parse command-line options
+        auto opts =
+            cl::sdk::parse_cli<cl::sdk::options::Diagnostic,
+                               cl::sdk::options::SingleDevice, BinariesOptions>(
+                argc, argv);
+        const auto& diag_opts = std::get<0>(opts);
+        const auto& dev_opts = std::get<1>(opts);
+        const auto& binaries_opts = std::get<2>(opts);
+
+        // Create context
+        cl::Context context = cl::sdk::get_context(dev_opts.triplet);
+
+        cl_int error;
+        std::vector<cl::Device> devices = context.getInfo<CL_CONTEXT_DEVICES>();
+
+        cl::CommandQueue queue{ context, devices[0],
+                                cl::QueueProperties::Profiling };
+
+        cl::Platform platform{
+            devices.at(0).getInfo<CL_DEVICE_PLATFORM>()
+        }; // https://github.com/KhronosGroup/OpenCL-CLHPP/issues/150
+
+        if (!diag_opts.quiet)
+        {
+            std::cout << "Selected platform: "
+                      << platform.getInfo<CL_PLATFORM_VENDOR>() << "\n"
+                      << "Selected device: "
+                      << devices.at(0).getInfo<CL_DEVICE_NAME>() << "\n"
+                      << std::endl;
+        }
+
+        /// Try to read binary
+        cl::Program::Binaries binaries =
+            cl::util::read_binary_files(devices, "Collatz", &error);
+
+        if (error != CL_SUCCESS)
```

Comment:
All samples are built with exception handling enabled. If error handling is implemented using `errHandler`, then in the presence of exceptions, trying to read the binaries which aren't on disk yet result in an exception. As such, compiling the binaries should be in a `catch(cl::util::Error&)` clause, so if any of the files did not exist on disk, then we compile from scratch and save.

Reply:
As a sidenote, the utility library could expose a kernel caching mechanism of some sort. Either using the current working directory, or some other more sophisticated storage mechanism. Other libraries roll their own kernel cache (clFFT for instance), but this is a recurring issue and could be captured by a dedicated utility. This is nothing to act upon now, but we should capture in an issue.

Reply:
Thanks for review!

I've tried to address your comments in the following commit.
I hope its better now.